### ISSUE TITLE
Decision log flushing unit tests hang

### ISSUE THREAD

**Issue by patrick-east** [open-policy-agent/opa#2792]
As seen here: https://pipelines.actions.githubusercontent.com/bRAwe4oNxdSVvPnva1R1AHyFMDMCFa4vxvrHwHAQpm4I22WpcW/_apis/pipelines/1/runs/644/signedlogcontent/11?urlExpires=2020-10-15T19%3A29%3A46.6818445Z&urlSigningMethod=HMACV1&urlSignature=DOQrOKbHddguwZkFZWI7gkAwG8DYXY1UeO4rh5siK%2FI%3D

Relavent bits:
```
2020-10-15T05:29:06.1016997Z ok  	github.com/open-policy-agent/opa/plugins	0.007s	coverage: 53.0% of statements
2020-10-15T05:29:07.1705719Z ok  	github.com/open-policy-agent/opa/plugins/bundle	0.062s	coverage: 85.0% of statements
2020-10-15T05:29:07.5484639Z ok  	github.com/open-policy-agent/opa/plugins/discovery	0.066s	coverage: 86.8% of statements
2020-10-15T05:39:08.9299840Z time="2020-10-15T05:29:09Z" level=info msg="Starting decision logger." plugin=decision_logs
2020-10-15T05:39:08.9301427Z time="2020-10-15T05:29:09Z" level=info msg="Log upload skipped." plugin=decision_logs
2020-10-15T05:39:08.9302885Z time="2020-10-15T05:29:09Z" level=info msg="Stopping decision logger." plugin=decision_logs
2020-10-15T05:39:08.9304256Z time="2020-10-15T05:29:09Z" level=info msg="Flushing decision logs." plugin=decision_logs
2020-10-15T05:39:08.9305295Z time="2020-10-15T05:29:09Z" level=info msg="All decisions in buffer uploaded." plugin=decision_logs
2020-10-15T05:39:08.9306247Z time="2020-10-15T05:29:09Z" level=info msg="Starting decision logger." plugin=decision_logs
2020-10-15T05:39:08.9307173Z time="2020-10-15T05:29:09Z" level=info msg="Stopping decision logger." plugin=decision_logs
2020-10-15T05:39:08.9308908Z time="2020-10-15T05:29:09Z" level=info msg="Flushing decision logs." plugin=decision_logs
2020-10-15T05:39:08.9310054Z time="2020-10-15T05:29:09Z" level=error msg="Log upload failed, server replied with HTTP 500." plugin=decision_logs
2020-10-15T05:39:08.9311488Z time="2020-10-15T05:29:09Z" level=info msg="Log upload skipped." plugin=decision_logs
2020-10-15T05:39:08.9312628Z time="2020-10-15T05:29:09Z" level=error msg="Graceful shutdown period ended with decisions possibly still in buffer." plugin=decision_logs
2020-10-15T05:39:08.9315131Z time="2020-10-15T05:29:09Z" level=error msg="Log upload failed: Post \"http://127.0.0.1:33279/logs\": context deadline exceeded." plugin=decision_logs
2020-10-15T05:39:08.9316635Z time="2020-10-15T05:29:10Z" level=error msg="Log upload failed: Post \"http://127.0.0.1:33279/logs\": context deadline exceeded." plugin=decision_logs
2020-10-15T05:39:08.9318012Z time="2020-10-15T05:29:11Z" level=error msg="Log upload failed: Post \"http://127.0.0.1:33279/logs\": context deadline exceeded." plugin=decision_logs
2020-10-15T05:39:08.9320285Z time="2020-10-15T05:29:12Z" level=error msg="Log upload failed: Post \"http://127.0.0.1:33279/logs\": context deadline exceeded." plugin=decision_logs
2020-10-15T05:39:08.9321764Z time="2020-10-15T05:29:13Z" level=error msg="Log upload failed: Post \"http://127.0.0.1:33279/logs\": context deadline exceeded." plugin=decision_logs
2020-10-15T05:39:08.9322723Z 2020/10/15 05:29:14 httptest.Server blocked in Close after 5 seconds, waiting for connections:
2020-10-15T05:39:08.9323386Z   *net.TCPConn 0xc00000e090 127.0.0.1:37542 in state active
2020-10-15T05:39:08.9324476Z time="2020-10-15T05:29:14Z" level=error msg="Log upload failed: Post \"http://127.0.0.1:33279/logs\": context deadline exceeded." plugin=decision_logs
2020-10-15T05:39:08.9325851Z time="2020-10-15T05:29:15Z" level=error msg="Log upload failed: Post \"http://127.0.0.1:33279/logs\": context deadline exceeded." plugin=decision_logs

< that repeats a whole bunch... >


2020-10-15T05:39:09.0235852Z time="2020-10-15T05:39:08Z" level=error msg="Log upload failed: Post \"http://127.0.0.1:33279/logs\": context deadline exceeded." plugin=decision_logs
2020-10-15T05:39:09.0236504Z coverage: 76.8% of statements
2020-10-15T05:39:09.0237000Z panic: test timed out after 10m0s
2020-10-15T05:39:09.0237255Z 
2020-10-15T05:39:09.0237522Z goroutine 1474 [running]:
2020-10-15T05:39:09.0238389Z testing.(*M).startAlarm.func1()
2020-10-15T05:39:09.0238844Z 	/usr/local/go/src/testing/testing.go:1524 +0xdf
2020-10-15T05:39:09.0239230Z created by time.goFunc
2020-10-15T05:39:09.0239590Z 	/usr/local/go/src/time/sleep.go:168 +0x44
2020-10-15T05:39:09.0239836Z 
2020-10-15T05:39:09.0240126Z goroutine 1 [chan receive, 9 minutes]:
2020-10-15T05:39:09.0240571Z testing.(*T).Run(0xc0002b0480, 0xb2848a, 0x3a, 0xb33da0, 0x4bca01)
2020-10-15T05:39:09.0241026Z 	/usr/local/go/src/testing/testing.go:1106 +0x399
2020-10-15T05:39:09.0241516Z testing.runTests.func1(0xc0001ae000)
2020-10-15T05:39:09.0242005Z 	/usr/local/go/src/testing/testing.go:1349 +0x78
2020-10-15T05:39:09.0242464Z testing.tRunner(0xc0001ae000, 0xc000093db8)
2020-10-15T05:39:09.0242915Z 	/usr/local/go/src/testing/testing.go:1054 +0xdc
2020-10-15T05:39:09.0243377Z testing.runTests(0xc00000c360, 0x10455c0, 0x17, 0x17, 0x0)
2020-10-15T05:39:09.0243822Z 	/usr/local/go/src/testing/testing.go:1347 +0x2a7
2020-10-15T05:39:09.0244205Z testing.(*M).Run(0xc00006e180, 0x0)
2020-10-15T05:39:09.0244593Z 	/usr/local/go/src/testing/testing.go:1264 +0x1b7
2020-10-15T05:39:09.0245472Z github.com/open-policy-agent/opa/plugins/logs.TestMain(0xc00006e180)
2020-10-15T05:39:09.0246035Z 	/src/plugins/logs/plugin_test.go:33 +0x4d
2020-10-15T05:39:09.0246368Z main.main()
2020-10-15T05:39:09.0246659Z 	_testmain.go:146 +0x1c1
2020-10-15T05:39:09.0246859Z 
2020-10-15T05:39:09.0247169Z goroutine 1448 [semacquire, 9 minutes]:
2020-10-15T05:39:09.0247660Z sync.runtime_Semacquire(0xc0000fa828)
2020-10-15T05:39:09.0248075Z 	/usr/local/go/src/runtime/sema.go:56 +0x42
2020-10-15T05:39:09.0248462Z sync.(*WaitGroup).Wait(0xc0000fa820)
2020-10-15T05:39:09.0249159Z 	/usr/local/go/src/sync/waitgroup.go:130 +0x64
2020-10-15T05:39:09.0249602Z net/http/httptest.(*Server).Close(0xc0000fa7e0)
2020-10-15T05:39:09.0250069Z 	/usr/local/go/src/net/http/httptest/server.go:249 +0x224
2020-10-15T05:39:09.0250946Z github.com/open-policy-agent/opa/plugins/logs.(*testServer).stop(0xc000159d80)
2020-10-15T05:39:09.0251522Z 	/src/plugins/logs/plugin_test.go:1206 +0x2f
2020-10-15T05:39:09.0253253Z github.com/open-policy-agent/opa/plugins/logs.TestPluginTerminatesImmediatelyAfterGracefulShutdownPeriod(0xc0002b0480)
2020-10-15T05:39:09.0254771Z 	/src/plugins/logs/plugin_test.go:555 +0x368
2020-10-15T05:39:09.0255197Z testing.tRunner(0xc0002b0480, 0xb33da0)
2020-10-15T05:39:09.0255622Z 	/usr/local/go/src/testing/testing.go:1054 +0xdc
2020-10-15T05:39:09.0256006Z created by testing.(*T).Run
2020-10-15T05:39:09.0256392Z 	/usr/local/go/src/testing/testing.go:1105 +0x372
2020-10-15T05:39:09.0256666Z 
2020-10-15T05:39:09.0256958Z goroutine 1457 [chan send, 9 minutes]:
2020-10-15T05:39:09.0258080Z github.com/open-policy-agent/opa/plugins/logs.(*testServer).handle(0xc000159d80, 0xbfb4a0, 0xc00033a0e0, 0xc000423200)
2020-10-15T05:39:09.0258748Z 	/src/plugins/logs/plugin_test.go:1197 +0x2a3
2020-10-15T05:39:09.0259810Z net/http.HandlerFunc.ServeHTTP(0xc00029c980, 0xbfb4a0, 0xc00033a0e0, 0xc000423200)
2020-10-15T05:39:09.0260453Z 	/usr/local/go/src/net/http/server.go:2041 +0x44
2020-10-15T05:39:09.0261084Z net/http.serverHandler.ServeHTTP(0xc000113340, 0xbfb4a0, 0xc00033a0e0, 0xc000423200)
2020-10-15T05:39:09.0261722Z 	/usr/local/go/src/net/http/server.go:2836 +0xa3
2020-10-15T05:39:09.0262165Z net/http.(*conn).serve(0xc0000d4c80, 0xbfd4a0, 0xc00006c9c0)
2020-10-15T05:39:09.0262607Z 	/usr/local/go/src/net/http/server.go:1924 +0x86c
2020-10-15T05:39:09.0262998Z created by net/http.(*Server).Serve
2020-10-15T05:39:09.0263370Z 	/usr/local/go/src/net/http/server.go:2962 +0x35c
2020-10-15T05:39:09.0263628Z 
2020-10-15T05:39:09.0263910Z goroutine 1452 [sleep]:
2020-10-15T05:39:09.0264239Z time.Sleep(0x3b9aca00)
2020-10-15T05:39:09.0264586Z 	/usr/local/go/src/runtime/time.go:188 +0xba
2020-10-15T05:39:09.0265859Z github.com/open-policy-agent/opa/plugins/logs.(*Plugin).flushDecisions.func1(0xc000196f20, 0xbfd4a0, 0xc00035c000, 0xc000028580)
2020-10-15T05:39:09.0266621Z 	/src/plugins/logs/plugin.go:431 +0x50
2020-10-15T05:39:09.0267772Z created by github.com/open-policy-agent/opa/plugins/logs.(*Plugin).flushDecisions
2020-10-15T05:39:09.0268563Z 	/src/plugins/logs/plugin.go:422 +0xc0
2020-10-15T05:39:09.0268948Z 
2020-10-15T05:39:09.0269214Z goroutine 1435 [sleep]:
2020-10-15T05:39:09.0269532Z time.Sleep(0x3b9aca00)
2020-10-15T05:39:09.0270645Z 	/usr/local/go/src/runtime/time.go:188 +0xba
2020-10-15T05:39:09.0272188Z github.com/open-policy-agent/opa/plugins/logs.(*Plugin).flushDecisions.func1(0xc0000c6420, 0xbfd4a0, 0xc00035c980, 0xc00013d3c0)
2020-10-15T05:39:09.0273093Z 	/src/plugins/logs/plugin.go:431 +0x50
2020-10-15T05:39:09.0274283Z created by github.com/open-policy-agent/opa/plugins/logs.(*Plugin).flushDecisions
2020-10-15T05:39:09.0275105Z 	/src/plugins/logs/plugin.go:422 +0xc0
2020-10-15T05:39:09.0276162Z FAIL	github.com/open-policy-agent/opa/plugins/logs	600.011s
```

**Comment by patrick-east** [comment]
The failing test was (I believe) added with https://github.com/open-policy-agent/opa/commit/3cd195c92544752b3d0efc447981297d2b82a732

**Comment by anderseknert** [comment]
Hmm, never saw this during development, on my own machine or in the build pipeline - of course it had to happen after merge 🙃 Trying to parse the logs pasted here (the link just gives me "Uri expired") it looks like the goroutine keeps going even after the context was canceled and the function returned. I'll work on a fix.

**Comment by patrick-east** [comment]
Re-opening, part of the fix is in #2803
